### PR TITLE
make size for multiple dimensions type stable

### PR DESCRIPTION
**Pull request by KristofferC**
cc @mbauman @timholy

### CODE REVIEW

**Comment by timholy** [comment]
Would the old definition work just as well if you added `@_inline_meta`?

**Comment by KristofferC** [comment]
With `size(x, d1::Integer, d2::Integer, dx::Integer...) = (@_inline_meta; tuple(size(x, d1), size(x, d2, dx...)...))`

I get

```jl
julia> @code_warntype size(rand(3,2,1), 2, 1)
Variables:
  #self#::Base.#size
  x::Array{Float64,3}
  d1::Int64
  d2::Int64
  dx::Tuple{}

Body:
  begin  # abstractarray.jl, line 24:
      return (Core._apply)(Base.tuple,(Core.tuple)((Base.arraysize)(x::Array{Float64,3},d1::Int64)::Int64)::Tuple{Int64},(Base.arraysize)(x::Array{Float64,3},d2::Int64)::Int64)::Tuple{Vararg{Any,N}}
  end::Tuple{Vararg{Any,N}}
```

which seems to indicate a negative.

**Comment by timholy** [comment]
Thanks for testing. I'm a bit surprised that doesn't work, I wonder why?

For reference, this more laborious form works:
```jl
julia> mysz(a, d::Integer) = size(a, d)
mysz (generic function with 1 method)

julia> @inline mysz(a, d1::Integer, d...) = _mysz((), a, d1, d...)
mysz (generic function with 2 methods)

julia> _mysz(out, a) = out
_mysz (generic function with 1 method)

julia> @inline _mysz(out, a, d1::Integer, d...) = _mysz((out..., mysz(a, d1)), a, d...)
_mysz (generic function with 2 methods)

julia> a = rand(Int, fill(3, 7)...);

julia> @code_warntype mysz(a, 2, 4, 6)
Variables:
  #self#::#mysz
  a::Array{Int64,7}
  d1::Int64
  d::Tuple{Int64,Int64}

Body:
  begin  # REPL[2], line 1:
      SSAValue(7) = (Core.getfield)(d::Tuple{Int64,Int64},1)::Int64
      SSAValue(8) = (Core.getfield)(d::Tuple{Int64,Int64},2)::Int64 # REPL[4], line 1:
      SSAValue(6) = (Base.arraysize)(a::Array{Int64,7},d1::Int64)::Int64 # REPL[4], line 1:
      SSAValue(4) = SSAValue(6)
      SSAValue(5) = (Base.arraysize)(a::Array{Int64,7},SSAValue(7))::Int64 # REPL[4], line 1:
      SSAValue(3) = (Core.tuple)(SSAValue(4),SSAValue(5),(Base.arraysize)(a::Array{Int64,7},SSAValue(8))::Int64)::Tuple{Int64,Int64,Int64} # REPL[3], line 1:
      return SSAValue(3)
  end::Tuple{Int64,Int64,Int64}
```

To clarify, I think this is merge-worthy, but I wonder if it's also an opportunity to fix an inference bug. I'm a little short on time for the next 48 hours, so I have no objections if someone merges this in the meantime.

**Comment by JeffBezanson** [comment]
I think the key part is `size(x, d2, dx...)...`. When `dx` is empty, the call to `size` returns an Int, which is then splatted, which only works because numbers are (controversially) iterable. It seems inference doesn't know that numbers are iterable.

**Comment by KristofferC** [comment]
So what to do here? Is there an advantage to @timholy's version? The disadvantage is, in my opinion, that is is harder to reason about that code, (four functions and recursive unrolling of arguments etc).

File: test/abstractarray.jl [hunk 1]
```diff
@@ -521,3 +521,5 @@ A = TSlowNIndexes(rand(2,2))
 @test_throws ErrorException A[1]
 @test A[1,1] == A.data[1]
 @test first(A) == A.data[1]
+
+@inferred size(rand(3,2,1), 2, 1)
```
Comment (timholy) [inline test/abstractarray.jl]: Maybe also add `@inferred size(rand(3,2,1), 2, 1, 3)` to ensure that you test when engaging a non-empty `Vararg`?

Comment (KristofferC) [inline test/abstractarray.jl]: Good point, added.

**Comment by KristofferC** [comment]
AppVeyor threw some GitError..

**Comment by JeffBezanson** [comment]
Ok I guess we should just merge this for now.

**Pull request merged**